### PR TITLE
[risk=low][RW-2707]notebook locking metadata endpoint

### PR DESCRIPTION
**Pull request by jmthibault79**
Also refactor storage.get() to return a better error

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 1]
```diff
@@ -690,13 +695,86 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
 
   @Override
   public ResponseEntity<ReadOnlyNotebookResponse> readOnlyNotebook(
-      String workspace, String workspaceName, String notebookName) {
+      String workspaceNamespace, String workspaceName, String notebookName) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: not really related - but I happened to notice this was wrong

File: api/src/main/java/org/pmiops/workbench/google/CloudStorageServiceImpl.java [hunk 2]
```diff
@@ -71,8 +73,17 @@ String getImagesBucketName() {
   }
 
   private String readToString(String bucketName, String objectPath) {
+    return new String(getBlob(bucketName, objectPath).getContent()).trim();
+  }
+
+  // wrapper for storage.get() which throws NotFoundException instead of NullPointerException
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/google/CloudStorageServiceImpl.java]: previously we used `storage.get().whatever()` without null checking.  it's still an error now, but a nicer one

File: api/src/main/java/org/pmiops/workbench/google/CloudStorageServiceImpl.java [hunk 3]
```diff
@@ -127,10 +138,15 @@ public GoogleCredential getDefaultServiceAccountCredentials() throws IOException
   }
 
   @Override
-  public JSONObject getFileAsJson(String bucketName, String fileName) throws IOException {
+  public JSONObject getFileAsJson(String bucketName, String fileName) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/google/CloudStorageServiceImpl.java]: unrelated misc change

File: api/src/test/java/org/pmiops/workbench/notebooks/NotebooksServiceImplTest.java [hunk 4]
```diff
@@ -52,7 +52,7 @@ User user() {
   @Autowired private CloudStorageService cloudStorageService;
 
   @Test
-  public void testGetReadOnlyHtml_basicContent() throws Exception {
+  public void testGetReadOnlyHtml_basicContent() {
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/notebooks/NotebooksServiceImplTest.java]: unrelated misc cleanup

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 5]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+    final String lastLockedUser = LOGGED_IN_USER_EMAIL;
+    final Long lockExpirationTime = Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli();
+
+    final Map<String, String> gcsMetadata =
+        new HashMap<String, String>() {
+          {
+            put("lockExpirationTime", lockExpirationTime.toString());
+            put(
+                "lastLockedBy",
+                WorkspacesController.notebookLockingEmailHash(BUCKET_NAME, lastLockedUser));
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: opt: Hardcoding the current output of this function might be preferable, so you can see a concrete input rather than depending on the implementation within the test. This can help in sanity checking, e.g. if Welder was returning the hash with capital hex letters and this was expected lowercase.

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 6]
```diff
@@ -690,13 +695,86 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
...
+  static String notebookLockingEmailHash(String bucket, String email) {
+    String toHash = String.format("%s:%s", bucket, email);
+    try {
+      MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+      byte[] hash = sha256.digest(toHash.getBytes(Charset.forName("UTF-8")));
+
+      // convert to printable hex text
+      StringBuilder sb = new StringBuilder();
+      for (byte b : hash) {
+        sb.append(String.format("%02x", b));
```
Comment (calbach) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: String.format is pretty slow and this would need to run 32*(num users in the acl); since we already pull in Guava - I might prefer an approach like this: https://stackoverflow.com/a/36477814

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 7]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace =
+        createFcWorkspace(testWorkspaceNamespace, testWorkspaceName, LOGGED_IN_USER_EMAIL);
+    fcWorkspace.setBucketName(BUCKET_NAME);
+    stubGetWorkspace(fcWorkspace, WorkspaceAccessLevel.OWNER);
+    stubFcGetWorkspaceACL(acl);
+
+    final String testNotebookPath = "notebooks/" + testNotebook;
+    doReturn(gcsMetadata).when(cloudStorageService).getMetadata(BUCKET_NAME, testNotebookPath);
+
+    // this is what I want to do, but it's broken somehow (never fails)
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: Looks like the paren is misplaced so this doesn't currently assert anything, should probably be:

```
assertThat(...getBody()).isEqualTo(expected);
```

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 8]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+          {
+            put("lockExpirationTime", lockExpirationTime.toString());
+            put(
+                "lastLockedBy",
+                WorkspacesController.notebookLockingEmailHash(BUCKET_NAME, lastLockedUser));
+            put("extraMetadata", "is not a problem");
+          }
+        };
+
+    // I can see that I have locked it myself, and when
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: incomplete?

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 9]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+    assertNotebookLockingMetadata(gcsMetadata, expectedResponse, workspaceACL);
+  }
+
+  @Test
+  public void testNotebookLockingMetadataUnknownUser() {
+    final String lastLockedUser = "a-stranger@fake-research-aou.org";
+    final Long lockExpirationTime = Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli();
+
+    final Map<String, String> gcsMetadata =
+        new HashMap<String, String>() {
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: Can this be `HashMap<>`, or does the anonymous class shenanigans below prevent that?

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 10]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+          {
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: This is really interesting - I've never seen this pattern used. I generally use an immutable map builder where I can... Does this anonymous class have an unfavorable side-effects? e.g. performance overhead or debug implications?

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java [hunk 11]
```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
...
+            put("lockExpirationTime", lockExpirationTime.toString());
+            // store directly in plaintext, to show that this does not work
+            put("lastLockedBy", lastLockedUser);
+            put("extraMetadata", "is not a problem");
+          }
+        };
+
+    // in case of accidentally storing the user email in plaintext
```
Comment (calbach) [inline api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: Nice test case

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 8 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: meaning "and when I locked it" - does this need clarification?

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 5 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: Agree that this is useful.  I added notebookLockingEmailHashTest() as an equivalent

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java (hunk 6 above)
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Thanks, updated

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 7 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: thanks, fixed

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 9 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: anon-class doesn't allow it.  removed regardless.

File: api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java (hunk 10 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java]: ImmutableMap.Builder looks better so I updated it to use that.

I should really have a look around Guava to see what else I'm missing.

File: api/src/main/resources/workbench.yaml [hunk 12]
```diff
@@ -1303,6 +1303,31 @@ paths:
...
+    get:
+      tags:
+        - workspaces
+      summary: Get the locking metadata for a notebook
+      description: |
+        Get the locking metadata for a notebook, if available.  If the notebook is not locked
+        (e.g. at first creation) this may be empty.  Additionally, the caller of this endpoint
+        may not have the appropriate permissions to determine who is locking this notebook.
+        When this is the case, the lastLockedBy field is set to UNKNOWN.
+      operationId: "notebookLockingMetadata"
```
Comment (calbach) [inline api/src/main/resources/workbench.yaml]: nit: most of our operation ids are verbs, since they translate into method names in swagger codegen. Of course, the endpoint immediately above this one violates this convention as well, so it could be classified as an existing issue.

File: api/src/main/resources/workbench.yaml [hunk 13]
```diff
@@ -3441,3 +3466,11 @@ definitions:
       notebookName:
         type: string
 
+  NotebookLockingMetadataResponse:
+    type: object
+    properties:
+      lastLockedBy:
+        type: string
+      lockExpirationTime:
+        type: integer
```
Comment (calbach) [inline api/src/main/resources/workbench.yaml]: It's probably worth adding a description here mentioning this is ms from the unix epoch (I assume it's ms).

File: api/src/main/resources/workbench.yaml (hunk 12 above)
Comment (jmthibault79) [reply api/src/main/resources/workbench.yaml]: ok, updated
